### PR TITLE
Allow optical elements to be inserted at arbitrary positions along FresnelOpticalSystem

### PR DESCRIPTION
**Pull request by sdwill**
Very straightforward- instead of appending an element to the end (default behavior), allow an optical element to be inserted somewhere else in a FresnelOpticalSystem.  This is useful for adding aberration planes to a system that was created elsewhere.

### CODE REVIEW

**Comment by mperrin** [comment]
Looks good. (And I wanted this functionality just the other day so this is timely. :-) We should add similar functionality to OpticalSysfem as well, too.

File: poppy/fresnel.py [hunk 1]
```diff
@@ -1066,16 +1066,27 @@ def add_optic(self, optic=None, distance=0.0 * u.m):
             Some optic
         distance : astropy.Quantity of dimension length
             separation distance of this optic relative to the prior optic in the system.
+        index : int
+            Index at which to insert the new optical element
+
         """
-        self.planes.append(optic)
-        self.distances.append(distance.to(u.m))
+        if index is None:
```
[the remark this answers is not in the document]
Comment (sdwill) [reply poppy/fresnel.py]: That is indeed more compact and readable- to be honest, I didn't think too much more about the initial implementation after I originally added it.  I'm fine with shortening this.

File: poppy/fresnel.py [hunk 2]
```diff
@@ -1066,16 +1066,27 @@ def add_optic(self, optic=None, distance=0.0 * u.m):
...
         if self.verbose:
             _log.info("Added optic: {0} after separation: {1:.2e} ".format(self.planes[-1].name, distance))
 
         return optic
 
-    @u.quantity_input(distance=u.m, pixelscale=u.micron/u.pixel)
-    def add_detector(self, pixelscale=10*u.micron/u.pixel, fov_pixels=10*u.pixel, distance=0.0 * u.m):
+    @u.quantity_input(distance=u.m, pixelscale=u.micron / u.pixel)
+    def add_detector(self, pixelscale=10 * u.micron / u.pixel, fov_pixels=10 * u.pixel,
+                     distance=0.0 * u.m):
```
[the remark this answers is not in the document]
Comment (sdwill) [reply poppy/fresnel.py]: I can't think of a common use case for placing a detector anywhere but the end of a beam train, so I would lean against that.

I would even go a step further and argue that for a Fresnel-type propagation where no distinctions are made between pupil and image planes, it makes more sense to use add_optic(Detector(...)), which then follows the same logic as any other optical element, covering the case here, and in my opinion is a bit less confusing.

[the remark this answers is not in the document]
Comment (mperrin) [reply poppy/fresnel.py]: > it makes more sense to use add_optic(Detector(...))

I agree with that. I've had similar thoughts occasionally about that bit of API, but haven't yet taken any steps to e.g. deprecate `add_detector` for removal in some future release.

**Pull request merged**